In [ ]:
from sklearn.ensemble import RandomForestClassifier,RandomForestRegressor
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder,OrdinalEncoder
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
df=pd.read_csv('loan_data.csv.xls')

In [ ]:
model=RandomForestClassifier(n_estimators=100)

In [ ]:
X = df.drop('loan_status', axis=1)
y = df['loan_status']

In [ ]:
df

In [ ]:
"Preprossed techinque used is onehotencoder"
x = df.drop(columns='loan_status')
y=df.loan_status


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

In [ ]:


preprocessing = ColumnTransformer(
    transformers=[
        ('onehot_encoder',
         OneHotEncoder(sparse_output=False, handle_unknown='ignore'),
         ['person_gender',
          'person_home_ownership',
          'loan_intent',
          'previous_loan_defaults_on_file']),

        ('ordinal',
         OrdinalEncoder(categories=[[
             'High School',
             'Associate',
             'Bachelor',
             'Master',
             'Doctorate'
         ]]),
         ['person_education'])
    ],
    remainder='passthrough'
)

conclude wheter is model overfit/under/good 
do pipeline as well
apply 2 preprocessing technique

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(random_state=42)

pipe = Pipeline([
    ('preprocessing', preprocessing),
    ('model', rf)
])

In [ ]:
pipe.fit(X_train, y_train)

In [ ]:
y_pred = pipe.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

In [45]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))




              precision    recall  f1-score   support

           0       0.94      0.97      0.95      6990
           1       0.89      0.77      0.83      2010

    accuracy                           0.93      9000
   macro avg       0.91      0.87      0.89      9000
weighted avg       0.93      0.93      0.93      9000



In [46]:
train_accuracy = pipe.score(X_train, y_train)
test_accuracy = pipe.score(X_test, y_test)

print("Training Accuracy:", train_accuracy)
print("Testing Accuracy:", test_accuracy)

Training Accuracy: 1.0
Testing Accuracy: 0.9283333333333333


The Random Forest model achieved 100% training accuracy and 92.83% testing accuracy. Since the training accuracy is significantly higher than the testing accuracy (difference ≈ 7.17%), the model shows slight overfitting. However, the testing accuracy is still high, indicating that the model generalizes reasonably well and performs effectively on unseen data.